File: dns/compare_dns_records.py
#!/usr/bin/env python3
"""
Compare two TOML files created by ``save_dns_records_as_toml.py``.
"""

import itertools
import sys

import toml


if __name__ == "__main__":
    try:
        saved_path = sys.argv[1]
        live_path = sys.argv[2]
    except IndexError:
        sys.exit(f"Usage: {__file__} SAVED_PATH LIVE_PATH")

    with open(saved_path) as in_file:
        saved_records = {k: set(v) for k, v in toml.load(in_file).items()}

    with open(live_path) as in_file:
        live_records = {k: set(v) for k, v in toml.load(in_file).items()}

    if saved_records == live_records:
        print("The DNS records match :D")
        sys.exit(0)
    else:
        print("The DNS records don't match D:", file=sys.stderr)

        for k in sorted(set(itertools.chain(saved_records, live_records))):
            if saved_records.get(k) != live_records.get(k):
                print("")
                print(f"-- {k}:")
                print(f"        saved: {saved_records.get(k)}")
                print(f"        life:  {live_records.get(k)}")

        sys.exit(1)

File: dns/compare_dns_records.rb
#!/usr/bin/env ruby
# Compare two YAML files created by ``save_dns_records_as_yaml.rb``
#
# This script takes two arguments: each a path to a YAML file that
# describes some DNS records.  The script then compares the records:
#
#   - If the DNS records are the same, it exits with code 0
#   - If the DNS records differ, it exits with code 1
#

require 'yaml'

unless ARGV.length == 2
  warn "Usage: #{$PROGRAM_NAME} SAVED_RECORDS_YAML LIVE_RECORDS_YAML"
  exit 1
end

saved_records = YAML.load_file(ARGV[0])
live_records = YAML.load_file(ARGV[1])

if saved_records == live_records
  puts 'The DNS records match 🥳'
  exit 0
else
  puts "The DNS records don't match! 😱"

  (saved_records.keys + live_records.keys).uniq.each do |k|
    next unless saved_records[k] != live_records[k]

    puts "#{k}:"
    puts " - saved: #{saved_records[k].inspect}"
    puts " - live:  #{live_records[k].inspect}"
  end

  exit 1
end
